core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+/*******************************************************************************
+ * Copyright (c) 2016 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+
+package org.eclipse.rdf4j.sail.inferencer.fc;
+
+import org.eclipse.rdf4j.common.iteration.Iterations;
+import org.eclipse.rdf4j.model.Resource;
+import org.eclipse.rdf4j.model.Statement;
+import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
+import org.eclipse.rdf4j.repository.Repository;
+import org.eclipse.rdf4j.repository.RepositoryConnection;
+import org.eclipse.rdf4j.repository.RepositoryResult;
+import org.eclipse.rdf4j.sail.NotifyingSail;
+import org.eclipse.rdf4j.sail.SailConflictException;
+import org.eclipse.rdf4j.sail.SailException;
+import org.eclipse.rdf4j.sail.inferencer.InferencerConnection;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.StampedLock;
+import java.util.stream.Collectors;
+
+/**
+ * @author Håvard Mikkelsen Ottestad
+ */
+
+public class ForwardChainingSchemaCachingRDFSInferencer extends AbstractForwardChainingInferencer {
+
+    // The schema, or null
+    Repository schema;
+
+    private static final Random random = new Random();
+
+    private StampedLock readWriteLock = new StampedLock();
+
+    // If false, the inferencer will skip some RDFS rules.
+    boolean useAllRdfsRules = true;
+
+    // Tbox cache
+    Collection<Resource> properties = new HashSet<>();
+    Collection<Resource> types = new HashSet<>();
+    Collection<Statement> subClassOfStatements = new ArrayList<>();
+    Collection<Statement> subPropertyOfStatements = new ArrayList<>();
+    Collection<Statement> rangeStatements = new ArrayList<>();
+    Collection<Statement> domainStatements = new ArrayList<>();
+
+    // Forward chained Tbox cache as lookup tables
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+
+    // The inferencer has been instantiated from another inferencer and shares it's schema with that one
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
```

Comment:
These changes are better done by calling .clear(), rather than changing the references. In multi-threaded situations if you want to reassign instance variables you would need "volatile".